# Exploring Ebay Car Sales Data

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

The dataset was originally [scraped](https://en.wikipedia.org/wiki/Web_scraping) and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data). A few modifications were made by [Dataquest](https://www.dataquest.io/) to the original dataset that was uploaded to Kaggle:

- The dataset was trimmed down to 50,000 data points from the full dataset
- The dataset was dirtied a bit to more closely resemble what could be expected from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with)

This was done to allow data scientists to put into use their data cleaning skills.

The aim of this project is to clean the data and analyze the included used car listings.

### Summary of Results
After analyzing the data, we reached the conclusion that the top 6 most listed car brands vary greatly in their prices. The Mercedes Benz cars were listed for an average price of about 8 800 whereas the Opel cars were listed for about 3 400. Their mileage was not a factor that affected the price of these brands.

Moreover, we observed that there were 3 times as many manual transmission cars than automatic transmission cars listed and that the mean price of automatic transmission cars in this dataset was more than twice as high than that of manual transmission cars.

Finally, we also observed that around 8% of the cars listed had unrepaired damage. The average price of damaged cars wass almost 4 times lower than that of undamaged cars.

For more details, please refer to the the full analysis below.

## Reading in the Dataset

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv',encoding='Latin-1') # UTF-8 raised an error

We can now visualize the first and last couple of rows:

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

The data dictionary provided with data is as follows:

- `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
- `name` - Name of the car.
- `seller` - Whether the seller is private or a dealer.
- `offerType` - The type of listing
- `price` - The price on the ad to sell the car.
- `abtest` - Whether the listing is included in an A/B test.
- `vehicleType` - The vehicle Type.
- `yearOfRegistration` - The year in which the car was first registered.
- `gearbox` - The transmission type.
- `powerPS` - The power of the car in PS.
- `model` - The car model name.
- `kilometer` - How many kilometers the car has driven.
- `monthOfRegistration` - The month in which the car was first registered.
- `fuelType` - What type of fuel the car uses.
- `brand` - The brand of the car.
- `notRepairedDamage` - If the car has a damage which is not yet repaired.
- `dateCreated` - The date on which the eBay listing was created.
- `nrOfPictures` - The number of pictures in the ad.
- `postalCode` - The postal code for the location of the vehicle.
- `lastSeenOnline` - When the crawler saw this ad last online.

Using the pandas `.info()` and `.head()` methods, we can get more information about the autos dataframe and its first few rows:

In [3]:
autos.info()

autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

At first glance, we notice that the dataset contains 20 columns, most of which are of `dtype` 'object', meaning strings.

There are several columns with missing values (never more than 20% null values), the one with the most missing values being 'notRepairedDamage'. This is understandable as there should only be a value if the car has a damage which is not yet repaired. Looking at the first couple of rows, we notice that most of the listings have entered 'nein' (no) in that field.

Some fields such as 'price' is of `dtype` 'object' and we notice that this is because of the '$' sign. This is something we will surely have to modify in order to better analyze the data.

We also notice that the column names might be difficult to reference because they use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case). We will have to convert the column names to snakecase and reword some of them based on the data dictionary to be more descriptive.

## Cleaning Column Names

In [4]:
autos.columns # Print the existing column names

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
# Make edits to column names - camelcase to snakecase
new_cols = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test','vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model','odometer', 'registration_month', 'fuel_type', 'brand','unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code','last_seen'] 

# Assign modified column names back to the columns attribute
autos.columns = new_cols

Other than the camelcase to snakecase conversions, other changes were made to ensure clarity in our column names:

- `yearOfRegistration` to `registration_year`
- `monthOfRegistration` to `registration_month`
- `notRepairedDamage` to `unrepaired_damage`
- `dateCreated` to `ad_created`

Let's look at the current state of the `autos` dataframe.

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Exploring the Dataset

We can now start doing some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for: 

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

To start off, we can use the `describe()` and `value_counts()` methods. 

In [7]:
autos.describe(include="all") # To get both categorical and numeric columns

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

#### Determining Columns to Drop

First, we can determine which columns have mostly one value and are candidates to be dropped:

- `seller`: only has 2 unique values, 4999 of which have a value of 'privat'
- `offer_type`: also only has 2 unique values, 4999 of which have a value of 'Angebot'
- `ab_test`: although it only has 2 unique values, there can really only be 2 possible values. We will not remove this one.
- `gearbox`: same as `ab_test`
- `unrepaired_damage`: needs more investigation before dropping, still unclear what this column contains

To explore the `unrepaired_damage` column, we can use the `value_counts()` method:

In [8]:
autos["unrepaired_damage"].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

It is clear that this column will be important to keep in order to determine whether cars are listed with a damage to be repaired or not.

#### Determining Columns to Clean

Next, we will check the columns that have numeric data stored as text that need to be cleaned:

- `price`: appears to be stored as text because of the '$' sign.
- `odometer`: appears to be stored as text because of the 'km' value

Before modifying these columns, we can explore them individually using the `head()` method:

In [9]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [10]:
autos["odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

The `price` column appears to contain text values such as '$', commas and spaces.

The `odometer` column appears to contain text values such as 'km', commas and spaces.

## Cleaning Dataset

We can now remove all non-numeric characters and convert the columns to a numeric dtype:

In [11]:
# Last method allows to convert to float values

autos["price"] = autos["price"].str.replace(' ', '').str.replace('$', '').str.replace(',', '').astype(float)      

autos["odometer"] = autos["odometer"].str.replace(' ', '').str.replace('km', '').str.replace(',', '').astype(float)    


To ensure that our changes do not confuse us in the future, we have to rename the `odometer` column name:

In [12]:
autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the `odometer_km` and `price` columns even more in depth. Using minimum and maximum values we will look for any values that look unrealistically high or low (outliers) that we might want to remove.

In [13]:
# Checking how many unique values there are
print(autos["price"].unique().shape)
print('\n')
print(autos["odometer_km"].unique().shape)

(2357,)


(13,)


There is quite a low number of unique values in the `odometer_km` column.

In [14]:
# Checking min/max/median/mean etc
print(autos["price"].describe())
print('\n')
print(autos["odometer_km"].describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


The `price` column seems to have extremely high maximum values and some '0' values which are most likely bad entries, whereas the `odometer_km` column seems to be within a normal range.

In [15]:
# Checking counts of unique values
print(autos["price"].value_counts().sort_index(ascending=True).head(n=10),'\n\n') # Showing from lowest to highest unique value counts
print(autos["price"].value_counts().sort_index(ascending=False).head(n=10),'\n\n') # Showing from highest to lowest unique value counts
print(autos["odometer_km"].value_counts().sort_index(ascending=True).head(),'\n\n')
print(autos["odometer_km"].value_counts().sort_index(ascending=False).head(),'\n\n')


0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
Name: price, dtype: int64 


99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
Name: price, dtype: int64 


5000.0     967
10000.0    264
20000.0    784
30000.0    789
40000.0    819
Name: odometer_km, dtype: int64 


150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
Name: odometer_km, dtype: int64 




#### Cleaning the 'Price' Column

We observe that the `price` column has **many** outliers. It will be important to clear them out in order to not skew our data unintentionally.

Before removing these outliers, we have to select objective values in order to define an acceptable price range that used cars would be sold at. Although we are unsure about the currency listed on our dataset (even though the dollar sign was used, we cannot be sure that these prices are not originally in euros), we can use a website such as the [Car Gurus website](https://www.cargurus.com/Cars/forsale) to determine an acceptable price range for our used cars. The lowest possible values seem to be set at `$1000` and the highest value at `$200,000`. To make sure we don't exclude listings that will considerably impact our analysis, we can define a minimum of `$500` and a maximum of `$500,000`.

We can now proceed to removing outliers.

In [16]:
# Assigning the cleaned dataset back to 'autos'
autos = autos[autos["price"].between(500,500000)]

# Exploring the top 10 rows
print('Top 10 Rows\n') 
autos.head(n=10)

Top 10 Rows



date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
12  2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   

    seller offer_type   price  ab_test vehicle_type  registration_year  \
0   privat    Angebot  5000.0  control          bus               2004   
1   privat    Angebot  8500.0  control    limousine               1997   
2   privat    Angebot  8990.0     test    limousine               2009   
3   privat    Angebot  4350.0  control   kleinwagen               2007   
4   privat    Angebot  1350.0     test        kombi               2003   
5   privat    Angebot  7900.0     test          bus               2006   
7   privat    Angebot  1990.0  control    limousine               1998   
9   privat    Angebot   590.0  control          bus               1997   
10  privat    Angebot   999.0     test          NaN               2017   
12  privat    Angebot  5299.0  control   kleinwagen               2010   

      gearbox  power_ps    model  odometer_km  registration_month fuel_type  \
0     manuell       158   andere     150000.0                   3       lpg   
1   automatik       286      7er     150000.0                   6    benzin   
2     manuell       102     golf      70000.0                   7    benzin   
3   automatik        71   fortwo      70000.0                   6    benzin   
4     manuell         0    focus     150000.0                   7    benzin   
5   automatik       150  voyager     150000.0                   4    diesel   
7     manuell        90     golf     150000.0                  12    diesel   
9     manuell        90   megane     150000.0                   7    benzin   
10    manuell        90      NaN     150000.0                   4    benzin   
12  automatik        71   fortwo      50000.0                   9    benzin   

         brand unrepaired_damage           ad_created  nr_of_pictures  \
0      peugeot              nein  2016-03-26 00:00:00               0   
1          bmw              nein  2016-04-04 00:00:00               0   
2   volkswagen              nein  2016-03-26 00:00:00               0   
3        smart              nein  2016-03-12 00:00:00               0   
4         ford              nein  2016-04-01 00:00:00               0   
5     chrysler               NaN  2016-03-21 00:00:00               0   
7   volkswagen              nein  2016-03-16 00:00:00               0   
9      renault              nein  2016-03-16 00:00:00               0   
10  volkswagen              nein  2016-03-14 00:00:00               0   
12       smart              nein  2016-03-31 00:00:00               0   

    postal_code            last_seen  
0         79588  2016-04-06 06:45:54  
1         71034  2016-04-06 14:45:08  
2         35394  2016-04-06 20:15:37  
3         33729  2016-03-15 03:16:28  
4         39218  2016-04-01 14:38:50  
5         22962  2016-04-06 09:45:21  
7         53474  2016-04-07 03:17:32  
9         15749  2016-04-06 10:46:35  
10        86157  2016-04-07 03:16:21  
12        34590  2016-04-06 14:17:52

In [17]:
# Exploring lowest unique values
print('Lowest Unique Value Counts\n') 
print(autos["price"].value_counts().sort_index(ascending=True).head(n=10))

Lowest Unique Value Counts

500.0    781
501.0      1
510.0      2
517.0      1
520.0      8
525.0      4
530.0      8
540.0      2
549.0     13
550.0    356
Name: price, dtype: int64


In [18]:
# Exploring highest unique values
print('Highest Unique Value Counts\n') 
print(autos["price"].value_counts().sort_index(ascending=False).head(n=10))

Highest Unique Value Counts

350000.0    1
345000.0    1
299000.0    1
295000.0    1
265000.0    1
259000.0    1
250000.0    1
220000.0    1
198000.0    1
197000.0    1
Name: price, dtype: int64


Taking a look at the data, we notice that the `price` column finally seems to be cleaned up properly.

#### Exploring the Date Columns

Let's now move on to the date columns and understand the date range the data covers. There are 5 columns that represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the `date_crawled`, `last_seen`, and `ad_created` columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values, so we can use methods like `describe()` to understand the distribution without any extra data processing.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [19]:
# Print first 5 rows of these columns
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

The first 10 string values in these columns represent the date. In order to understand the date range, we can extract just the date values, use `value_counts()` to generate a distribution, and then sort by the index.

To select the first 10 characters in each column, we can use `Series.str[:10]`.

#### Distribution of Crawl Dates

In [20]:
# Combining methods to view distribution of crawl dates
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index() # sort_index allows to view from earliest to latest date  


2016-03-05    0.025567
2016-03-06    0.014125
2016-03-07    0.036189
2016-03-08    0.033173
2016-03-09    0.032907
2016-03-10    0.032707
2016-03-11    0.033018
2016-03-12    0.037320
2016-03-13    0.015522
2016-03-14    0.036300
2016-03-15    0.034016
2016-03-16    0.029359
2016-03-17    0.031155
2016-03-18    0.012883
2016-03-19    0.034747
2016-03-20    0.038073
2016-03-21    0.037741
2016-03-22    0.033018
2016-03-23    0.032397
2016-03-24    0.028982
2016-03-25    0.031089
2016-03-26    0.032641
2016-03-27    0.031177
2016-03-28    0.034836
2016-03-29    0.033262
2016-03-30    0.033328
2016-03-31    0.031665
2016-04-01    0.033905
2016-04-02    0.035767
2016-04-03    0.038827
2016-04-04    0.036610
2016-04-05    0.013172
2016-04-06    0.003171
2016-04-07    0.001353
Name: date_crawled, dtype: float64

The crawl dates seem to be evenly distributed (between 1% and 4% each) until the last 2 listed crawl dates. The programmer allowed their bot to crawl the data daily during a full month, from March 2016 to April 2016.

#### Distribution of Ad Posting Dates

In [21]:
# Combining methods to view distribution of post dates
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index() # sort_index allows to view from earliest to latest date  


2015-06-11    0.000022
2015-08-10    0.000022
2015-09-09    0.000022
2015-11-10    0.000022
2015-12-05    0.000022
2015-12-30    0.000022
2016-01-03    0.000022
2016-01-07    0.000022
2016-01-10    0.000044
2016-01-13    0.000022
2016-01-14    0.000022
2016-01-16    0.000022
2016-01-22    0.000022
2016-01-27    0.000067
2016-01-29    0.000022
2016-02-01    0.000022
2016-02-02    0.000044
2016-02-05    0.000044
2016-02-07    0.000022
2016-02-08    0.000022
2016-02-09    0.000022
2016-02-11    0.000022
2016-02-12    0.000044
2016-02-14    0.000044
2016-02-16    0.000022
2016-02-17    0.000022
2016-02-18    0.000044
2016-02-19    0.000067
2016-02-20    0.000044
2016-02-21    0.000044
                ...   
2016-03-09    0.032996
2016-03-10    0.032441
2016-03-11    0.033328
2016-03-12    0.037098
2016-03-13    0.016963
2016-03-14    0.034880
2016-03-15    0.033794
2016-03-16    0.029847
2016-03-17    0.030822
2016-03-18    0.013504
2016-03-19    0.033616
2016-03-20    0.038207
2016-03-21 

The ads crawled were created between June 2015 and April 2016. The majority of them seem to have been created during the crawling period. We can check this using a variation of our method chaining:

In [22]:
# Combining methods to view distribution of post dates
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).head(n=10) # sort_index allows to view from earliest to latest date  


2016-04-03    0.039049
2016-03-20    0.038207
2016-03-21    0.037963
2016-03-12    0.037098
2016-04-04    0.036987
2016-04-02    0.035457
2016-03-07    0.034903
2016-03-28    0.034880
2016-03-14    0.034880
2016-04-01    0.033838
Name: ad_created, dtype: float64

This indeed appears to be the case. Most of the ads that were crawled by the bot were posted during the crawling period.

#### Distribution of Last Seen Column

In [23]:
# Combining methods to view distribution of last seen dates
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index() # sort_index allows to view from earliest to latest date  


2016-03-05    0.001087
2016-03-06    0.004169
2016-03-07    0.005211
2016-03-08    0.007007
2016-03-09    0.009468
2016-03-10    0.010289
2016-03-11    0.012041
2016-03-12    0.023904
2016-03-13    0.008870
2016-03-14    0.012285
2016-03-15    0.015677
2016-03-16    0.016165
2016-03-17    0.027674
2016-03-18    0.007406
2016-03-19    0.015411
2016-03-20    0.020423
2016-03-21    0.020667
2016-03-22    0.021243
2016-03-23    0.018405
2016-03-24    0.019536
2016-03-25    0.018582
2016-03-26    0.016476
2016-03-27    0.015456
2016-03-28    0.020534
2016-03-29    0.021354
2016-03-30    0.024148
2016-03-31    0.023438
2016-04-01    0.022862
2016-04-02    0.024880
2016-04-03    0.024946
2016-04-04    0.024303
2016-04-05    0.126616
2016-04-06    0.225314
2016-04-07    0.134155
Name: last_seen, dtype: float64

The `last_seen` column indicates when the crawler last saw the ad online. Understandably, ads posted during most recent dates of crawling period constitute the majority of the ads last seen online.

#### Distribution of Registration Year

In [24]:
autos["registration_year"].describe()

count    45097.000000
mean      2005.064173
std         89.652017
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

We immediately observe that there are outliers in this column as a car may not have possibly been registered in the year 1000 or 9999. On the other hand, we observe that cars posted on this classifield were on average registered around 2005.

#### Cleaning Registration Year Column

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [25]:
# Counting values lower than 1900 and higher than 2016
autos[(autos["registration_year"] <1900) | (autos["registration_year"] >2016)] # Using pandas OR statement to include both
      

date_crawled                                               name  \
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
84     2016-03-27 19:52:54                                     Renault_twingo   
113    2016-04-03 14:58:29                              Golf_4_Anfaenger_auto   
164    2016-03-13 20:39:16  Opel_Meriva__nur_76000_Km__unfallfrei__scheckh...   
197    2016-04-05 10:36:24                              VW_Polo_9N_an_Bastler   
253    2016-03-27 13:25:18             Ford_mondeo_Gas_anlage_mit_TÜV_04.2017   
348    2016-03-17 20:58:24  VW_Beetle_1.8Turbo_mit_Vollausstattung_und_seh...   
390    2016-03-25 12:59:06   Fiat_Bertone_X_1_9__X_1/9__X19__X_19__X1_9__X_19   
438    2016-03-10 20:36:25                VW_Golf_5_V_1.6_2004_Klima_Tempomat   
453    2016-03-28 13:51:12                                         Armee_Jeep   
454    2016-03-25 09:57:03    Ford_Fiesta_Servo_D3_Kat_86€_Steuern_65.000_tkm   
457    2016-03-15 17:48:39  Mercedes_Benz_W203_C200_Kombi_Kompressor_c_kla...   
477    2016-03-31 08:53:34                                         Citroën_C4   
545    2016-03-22 18:50:53                                      Bmw_x5_/__E53   
548    2016-03-10 17:44:57            Peugeot_206_1.9_Hdi_4_Tueren_Tuev_02.17   
557    2016-03-16 03:36:34                           Smart_Fortwo_Passion_MHD   
568    2016-03-18 09:48:44             PRIMA_ANFÄNGERAUTO_smart_fortwo_cabrio   
577    2016-04-03 14:49:09                                  Volkswagen_Golf_3   
582    2016-03-31 11:57:46                                    V10_TDI_TOUAREG   
654    2016-04-01 15:48:34              Ich_verkaufe_meinen_gepflegten_daewoo   
740    2016-04-02 19:49:46                            Seat_Leon_zu_verkaufen!   
754    2016-03-26 13:47:56                         Ford_Ka_Kleinwagen_BJ_2008   
758    2016-03-12 18:42:33                            Honda_Civic_mit_TÜV_!!!   
765    2016-03-26 11:52:53               BMW_3er_Limousine_mit_1_Jahr_TÜV!!!!   
802    2016-03-19 11:37:23                 Lada_mit_wenig_km_neuem_Tuev_bj_08   
871    2016-04-04 19:38:38                                   Citroen_C2_60_PS   
889    2016-03-12 16:52:19                          Verkaufe_Opel_Corsa_C_1_2   
910    2016-03-15 17:52:50                         Renault_Kangoo_bj.2007_LPG   
918    2016-03-15 15:53:56                              BMW_114i_Sondermodell   
...                    ...                                                ...   
49075  2016-03-18 12:47:25           VW_Golf_2_GT_TÜV_neu_GTi_G60_16V_Special   
49076  2016-03-25 21:39:21                             Omega_B_2.0_16v_136_Ps   
49108  2016-03-14 22:37:20                  TDI_20_L_DIESEL140_PS..AUTOMATICK   
49130  2016-03-09 18:46:05                                      Renault_Modus   
49153  2016-03-12 01:36:59                                        Corsa_c20xe   
49178  2016-03-24 19:58:43       Mazda_323_F_TÜV__neu_2te_Hand_viele_Neuteile   
49185  2016-03-28 11:45:31              Alfa_Romeo_156_Sportwagon_1_8_T_Spark   
49218  2016-04-01 19:48:06          Ford_Fiesta_1.3_4/5_Tuerer_16_Monate_Tuev   
49245  2016-03-26 08:32:30                         Audi_A3_1.8T_tausch_moegl.   
49259  2016-03-19 18:45:05                  Volkswagen_Lupo_TÜV_06.2017_Klima   
49261  2016-03-10 12:51:40                                        Toyota_aygo   
49262  2016-03-26 18:31:21                                        Opel_vectra   
49266  2016-03-10 00:54:09  Toyota_Corolla_Verso_7_Sitzer_104.000km_2_2_Di...   
49283  2016-03-15 18:38:53                                         Citroen_HY   
49343  2016-03-12 11:56:55                                   bmw_320i_touring   
49347  2016-03-21 20:41:09                         Honda_civic_mit_TÜV_3/2017   
49354  2016-04-05 10:31:29                        Bmw_e39_523i_mit_neuem_Tuev   
49389  2016-03-10 07:54:10                   

There appears to be 1774 ads posted that have an incorrect registration year according to our criteria. We can proceed to remove these postings as they will incorrectly influence our analysis.

In [26]:
# Attributing to autos all rows that have a registration year between 1900 and 2016 (included)
autos = autos[(autos["registration_year"] >= 1900) & (autos["registration_year"] <= 2016)] # Using pandas AND statement to make sure both conditions are met

We can now properly calculate thte distribution og the remaining values:

In [27]:
autos["registration_year"].value_counts(normalize=True)

2005    0.066108
2000    0.062669
2004    0.061930
2006    0.061538
2003    0.061330
1999    0.059368
2001    0.058168
2002    0.055836
2007    0.052397
2008    0.050920
2009    0.047965
1998    0.046765
2011    0.037324
2010    0.036609
1997    0.034901
2012    0.030169
1996    0.024744
2016    0.021928
1995    0.019851
2013    0.018397
2014    0.015050
1994    0.011541
2015    0.008402
1993    0.007848
1992    0.007040
1991    0.006763
1990    0.006325
1989    0.003578
1988    0.002978
1985    0.001962
          ...   
1966    0.000485
1960    0.000439
1969    0.000439
1975    0.000415
1965    0.000392
1964    0.000254
1963    0.000185
1959    0.000138
1961    0.000138
1958    0.000092
1956    0.000092
1962    0.000092
1937    0.000092
1951    0.000046
1955    0.000046
1954    0.000046
1910    0.000046
1941    0.000046
1957    0.000046
1934    0.000046
1953    0.000023
1950    0.000023
1927    0.000023
1929    0.000023
1931    0.000023
1948    0.000023
1938    0.000023
1939    0.0000

The most common registration year appears to be 2005 (6.6%). There are a total of 78 unique registration years, several of which are from the early 1900s.

## Exploring Price by Brand

Because we are working with data on cars, it is natural to explore variations across different car brands. We will therefore aggregate the data by brand to explore the prices of the cars listed on the classifield.

The first step is to explore the unique values in the `brand` column:

In [28]:
# Checking unique value counts
print(autos["brand"].value_counts())

# Checking number of unique values
len(autos["brand"].value_counts())


volkswagen        9180
bmw               4983
mercedes_benz     4430
opel              4347
audi              3921
ford              2823
renault           1902
peugeot           1272
fiat              1019
seat               775
skoda              743
smart              655
mazda              650
nissan             650
citroen            613
toyota             581
hyundai            440
sonstige_autos     420
volvo              406
mini               406
honda              338
mitsubishi         335
kia                318
alfa_romeo         287
porsche            278
chevrolet          261
suzuki             256
chrysler           156
dacia              123
jeep               106
land_rover          98
daihatsu            94
subaru              86
saab                72
jaguar              70
daewoo              58
rover               54
trabant             48
lancia              43
lada                26
Name: brand, dtype: int64


40

There are a total of 40 unique brands in the dataset. For the sake of our analysis, we will work with the top 20 brands in the dataset. At first glance, it is evident that the top 3 brands are German brands, as the dataset was pulled from a German website. We can take a look at the top 20 brands:

In [29]:
# Checking top 20 and assigning to variable
brand_top20 = autos["brand"].value_counts().head(n=20).index 

print(*brand_top20, sep='\n')

volkswagen
bmw
mercedes_benz
opel
audi
ford
renault
peugeot
fiat
seat
skoda
smart
mazda
nissan
citroen
toyota
hyundai
sonstige_autos
volvo
mini


We will now aggregate the data on these brands:

In [30]:
# Creating empty dictionary to hold aggregate data
brand_mean_price = {}

for brand in brand_top20: # Looping through the top 20 index created
    key = autos[autos["brand"] == brand] # Creating a subset of the dataframe that contains only the unique brand
    val = key["price"].mean() # Calculating the average price for that brand
    brand_mean_price[brand] = round(val, 2) # Creating new key in the dictionary for the brand in question and assigning the average price rounded to 2 decimal places as a value for that key
    
for k,v in brand_mean_price.items(): # Printing dictionary line by line
    print(k,v)
    

renault 2819.06
smart 3611.04
fiat 3256.15
audi 9613.65
sonstige_autos 13439.73
nissan 5175.9
volkswagen 5783.62
volvo 5189.42
bmw 8582.26
skoda 6558.12
toyota 5266.97
mazda 4459.27
mini 10691.06
ford 4291.67
opel 3394.04
mercedes_benz 8766.9
hyundai 5686.94
peugeot 3360.92
seat 4810.88
citroen 4012.83


At first glance, we observe that Sonstige Autos, Mini and Audi have the highest average price. Furthermore, the prices range from 3 000 to 14 000. 

When comparing back to the top 6 most popular brands, we see that there is a distinct price gap:

- Audi, BMW and Mercedes Benz are more expensive
- Ford and Opel are less expensive
- Volkswagen is in between

For the top 6 brands, we can use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. 

#### Comparing Average Price and Average Mileage for Top 6 Brands

While a natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- it's difficult to compare more than two aggregate series objects if we want to extend to more columns
- we can't compare more than a few rows from each series object
- we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly.

Using a similar method as previously used, we can create a dictionary for mean mileage per brand.

In [31]:
# Creating empty dictionary to hold aggregate data
brand_mean_mileage = {}

for brand in brand_top20: # Looping through the top 20 index created
    key = autos[autos["brand"] == brand] # Creating a subset of the dataframe that contains only the unique brand
    val = key["odometer_km"].mean() # Calculating the average mileage for that brand
    brand_mean_mileage[brand] = round(val, 2) # Creating new key in the dictionary for the brand in question and assigning the average mileage rounded to 2 decimal places as a value for that key


We can then convert both the `brand_mean_price` and `brand_mean_mileage` dictionaries into series objects using the [series constructor](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html).

In [32]:
# Converting into series objects
bmp_series = pd.Series(brand_mean_price) # Mean price series

bmm_series = pd.Series(brand_mean_mileage) # Mean mileage series

We can then combine both series objects into a single dataframe using the [dataframe constructor](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) and start comparing the average mileage and prices of each brand.

In [33]:
# Creating dataframe from mean price series object first using dataframe constructor
price_mileage_dataframe = pd.DataFrame(bmp_series, columns=['mean_price']) # Naming column mean_price

# Assign mean mileage series object as new column in the dataframe
price_mileage_dataframe['mean_mileage'] = bmm_series

# Printing top 6 brands we are interested in
price_mileage_dataframe.loc[['volkswagen','bmw','mercedes_benz','opel','audi','ford']]


mean_price  mean_mileage
volkswagen        5783.62     128234.75
bmw               8582.26     132865.74
mercedes_benz     8766.90     131019.19
opel              3394.04     128012.42
audi              9613.65     128909.72
ford              4291.67     123494.51

When comparing the top 6 most common car brands posted on the classifield, we notice that although there is quite an important difference between each of their average price, their average mileage seem to be within a tight range (12 300 km to 13 300 km).

## Exploring Manual Versus Automatic Cars

We can continue exploring the dataset by looking at the `gearbox` column next. For example, we can check which type of gearbox is most common in this classifield.

In [34]:
autos['gearbox'].head()

0      manuell
1    automatik
2      manuell
3    automatik
4      manuell
Name: gearbox, dtype: object

We observe that the entries in this column are in German. In order to better analyze our data, we can translate them into English.

In [35]:
# Check all unique values
autos['gearbox'].value_counts(dropna=False) # Include NaN entries

manuell      32058
automatik     9661
NaN           1604
Name: gearbox, dtype: int64

Although the translations can be quite intuitive, it is important to carry through with our procedure and translate the German 'automatik' to the English 'automatic' and the German 'manuell' to the English 'manual'.

In [36]:
# Since there are only 2 unique values, we can chain the str.replace method directly without passing through a dictionary
autos['gearbox'] = autos['gearbox'].str.replace('automatik', 'automatic').str.replace('manuell','manual')

# Check unique values once translated
autos['gearbox'].value_counts(dropna=False) # Include NaN entries

manual       32058
automatic     9661
NaN           1604
Name: gearbox, dtype: int64

We notice that there are more than 3 times as many manual transmission cars as there are automatic transmission cars. The next thing we could check is the difference in the price between manual and automatic transmission cars.

In [37]:
# Assigning manual cars to a separate dataframe
manual_autos = autos[autos['gearbox'] == 'manual']

# Assigning automatic cars to a separate dataframe
automatic_autos = autos[autos['gearbox'] == 'automatic']

# Checking difference between manual and automatic car price
print('Manual Cars: \n', manual_autos['price'].describe(), '\n')

print('Automatic Cars: \n', automatic_autos['price'].describe())

Manual Cars: 
 count     32058.000000
mean       5083.818672
std        7162.541282
min         500.000000
25%        1350.000000
50%        2990.000000
75%        6500.000000
max      350000.000000
Name: price, dtype: float64 

Automatic Cars: 
 count      9661.000000
mean      11188.759342
std       13377.690585
min         500.000000
25%        3000.000000
50%        7490.000000
75%       15000.000000
max      345000.000000
Name: price, dtype: float64


We observe that the mean price of automatic transmission cars in this dataset is more than twice as high than that of manual transmission cars. This can be related to 2 factors:

1. Manual transmission cars are generally cheaper than automatic transmission cars, whether bought new or used.
2. There are significantly more manual transmission cars listed for sale than there are automatic transmission cars on this classifield. As a rule of thumb, a higher offer usually reduces the cost of buying a product.

## Exploring Damaged Versus Non-Damaged Cars

Next, we can see if there is a significant difference between the price of cars listed with a damage versus those listed without a damage.

In [38]:
# Exploring the damaged column
autos['unrepaired_damage'].value_counts(normalize=True, dropna=False)

nein    0.761535
NaN     0.156199
ja      0.082266
Name: unrepaired_damage, dtype: float64

We observe that there is over 15% of listings that did not disclaim whether there was any unrepaired damage to the car. We can continue the analysis by excluding these listings. As it was the case previously, the entries in this column are in German, however translating these two words is not necessary for our analysis.

We can now check the difference in price between damaged and undamged cars.

In [39]:
# Assigning undamaged cars to a separate dataframe
undamaged_autos = autos[autos['unrepaired_damage'] == 'nein']

# Assigning damaged cars to a separate dataframe
damaged_autos = autos[autos['unrepaired_damage'] == 'ja']

# Checking difference between manual and automatic car price
print('Undamaged Cars: \n', undamaged_autos['price'].describe(), '\n')

print('Damaged Cars: \n', damaged_autos['price'].describe())

Undamaged Cars: 
 count     32992.00000
mean       7339.22848
std       10145.65371
min         500.00000
25%        1950.00000
50%        4300.00000
75%        9300.00000
max      350000.00000
Name: price, dtype: float64 

Damaged Cars: 
 count     3564.000000
mean      2779.255331
std       3850.159721
min        500.000000
25%        800.000000
50%       1449.500000
75%       3200.000000
max      44200.000000
Name: price, dtype: float64


We observe that the average price of damaged cars is almost 4 times lower than that of undamaged cars. We also observe that the highest priced damaged car costs 44 200, whereas the highest priced undamaged car costs 350 000.

Even though this might seem logical, this means that overall, damaged cars are listed for a much lower price than their undamaged couterparts.

## Conclusion

The goal of this project was to explore and analyze a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

We reached the conclusion that the top 6 most listed car brands vary greatly in their prices. The Mercedes Benz cars were listed for an average price of about 8 800 whereas the Opel cars were listed for about 3 400. Their mileage was not a factor that affected the price of these brands.

Moreover, we observed that there were 3 times as many manual transmission cars than automatic transmission cars listed and that the mean price of automatic transmission cars in this dataset was more than twice as high than that of manual transmission cars.

Finally, we also observed that around 8% of the cars listed had unrepaired damage. The average price of damaged cars wass almost 4 times lower than that of undamaged cars.